controllers/database_controller.go

```diff
@@ -0,0 +1,282 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/sethvargo/go-password/password"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/strings/slices"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	stardogv1beta1 "github.com/vshn/stardog-userrole-operator/api/v1beta1"
+	"github.com/vshn/stardog-userrole-operator/pkg/stardogapi"
+
+	types "k8s.io/apimachinery/pkg/types"
+)
+
+// DatabaseReconciler reconciles a Database object
+type DatabaseReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+type User struct {
+	name     string
+	password string
+}
+
+//+kubebuilder:rbac:groups=stardog.vshn.ch,resources=databases,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=stardog.vshn.ch,resources=databases/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Database object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	r.Log = r.Log.WithValues("database", req.NamespacedName)
+	// TODO track connected external resources in status
+
+	database := &stardogv1beta1.Database{}
+	err := r.Get(ctx, req.NamespacedName, database)
+	if err != nil {
+		r.Log.Error(err, "error getting database")
+		return ctrl.Result{}, err
+	}
+
+	for _, instanceRef := range database.Spec.Instances {
```

Comment:
Is there a reason a database can reference multiple instances?

I'd strongly prefer only referencing one instance per database.

Reply:
The idea is that one only has to create a single object providing the name of a database to create it in all environments (refs) (also see https://wiki.vshn.net/x/T4WHE).

Reply:
This could also be solved with a `DatabaseTemplate` (name not final) resource creating and owning `Database` resources.  Which would decrease the complexity of this controller.

With the current way of instance lists you'd also need to track old list vs new list and react to changes there. What happens on rename and delete? Also made harder with this list: ownership and partial failure.

Reply:
`DatabaseSet` ?